### PR TITLE
improves the concat operation cases happening in yolo

### PR DESCRIPTION
**Pull request by quickwritereader**
## What changes were proposed in this pull request?

improves the concat operation cases happening in yolo

## How was this patch tested?

unit test

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [x] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.

### CODE REVIEW

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 1]
```diff
@@ -127,76 +127,142 @@ namespace sd {
...
+ */
+int isShapeExtendedWithOnes( const NDArray &input, int axis){
+   bool isAllOne=true;
+   auto shapes = shape::shapeOf(input.shapeInfo());
+   auto rank = input.rankOf();
+   if(rank>axis){
+
+      if(input.ordering()=='c'){
+        //check befor axis
+        for(int i=0; i<axis; i++){
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp]: Fix up formatting. Spaces between i = 0 and i < axis etc

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 2]
```diff
@@ -127,76 +127,142 @@ namespace sd {
...
-      ((axis == 0 && output.ordering() == 'c') || (axis == output.rankOf() - 1 && output.ordering() == 'f')) &&
-      output.ews() == 1;
-
-  if (luckCase1) {
-    for (sd::Unsigned i = 0; i < numOfInArrs; ++i) {
-      luckCase1 &= inArrs[i]->ordering() == output.ordering() && inArrs[i]->ews() == 1;
-      if (!luckCase1) break;
-    }
+  bool shapeExtendedWithOnes = isShapeExtendedWithOnes(output, axis);
+  bool followEws1 = output.ews()==1;
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp]: Fix spacing as above.

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 3]
```diff
@@ -127,76 +127,142 @@ namespace sd {
...
 
-  if (luckCase1) {  // for example {1,10} + {2,10} + {3,10} = {6, 10} order c; or {10,1} + {10,2} + {10,3} = {10, 6}
-                    // order f
-
-    T *z = zBuff;
-    for (sd::Unsigned i = 0; i < numOfInArrs; ++i) {
-      const auto memAmountToCopy = inArrs[i]->lengthOf();
-      memcpy(z, inArrs[i]->bufferAsT<T>(), memAmountToCopy * sizeofT);
-      z += memAmountToCopy;
+  bool copyCaseEws1 = followEws1 & matchesOutputOrdering;
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp:181]: Add some docs explaining the performance increase here.

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 4]
```diff
@@ -127,76 +127,142 @@ namespace sd {
...
+    for (int i = 0; i <numOfInArrs; i++){
+      zPtrList.push_back(z);
+      z += inArrs[i]->lengthOf();
     }
+    auto func = [&inArrs, &zPtrList](uint64_t thread_id, int64_t start, int64_t stop, int64_t increment) -> void {
+      for (int i = start; i < stop; ++i) {
+        const auto memAmountToCopy = inArrs[i]->lengthOf();
+        const auto inputPtr = inArrs[i]->bufferAsT<T>();
+  #if defined(__NEC__)
+        auto zPtr = zPtrList[i];
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp:203]: Spacing

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 5]
```diff
@@ -127,76 +127,142 @@ namespace sd {
...
-  // all inputs arrays and output array
+    output.assign(inArrs[0]);
+    return;
+  }
 
-  //     const auto zStep = strideOverContigAxis(axis, output.getShapeInfo());
+  if(copyCaseEws1 && output.ordering()=='c'){
 
-  //     for (sd::Unsigned i = 0; i < output.lengthOf() / output.sizeAt(axis); ++i) {
+    int times=1;
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp]: Spacing.

**Review by agibsonccc — CHANGES_REQUESTED** [review]

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 6]
```diff
@@ -127,76 +127,130 @@ namespace sd {
...
-  //     }
+    std::vector<InputArgsCase2<T>> inputArgs;
+    for (int i = 0; i < numOfInArrs; i++) {
+      InputArgsCase2<T> input = {inArrs[i]->bufferAsT<T>(), inArrs[i]->lengthOf() / times};
+      inputArgs.push_back(input);
+    }
 
-  //     return;
-  // }
+    auto func = [&inputArgs, z, totalCopySize](uint64_t thread_id, int64_t start, int64_t stop,
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp:248]: Shouldn't these be sd::LongType?

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 7]
```diff
@@ -127,76 +127,130 @@ namespace sd {
...
+  if (numOfInArrs < 2) {
+    output.assign(inArrs[0]);
+    return;
+  }
 
-  //         T* z = zBuff + zStep * i;
+  if (copyCaseEws1 && output.ordering() == 'c') {
+    int times = 1;
+    auto shapes = shape::shapeOf(output.shapeInfo());
+    assert(axis < rank);
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp]: Add a message to this describing the problem. Random errors without messages are VERY jarring especially from java.

File: libnd4j/include/ops/impl/specials_single.hpp [hunk 8]
```diff
@@ -127,76 +127,130 @@ namespace sd {
...
+ * @return int
+ */
+int isShapeExtendedWithOnes(const NDArray &input, int axis) {
+  bool isAllOne = true;
+  auto shapes = shape::shapeOf(input.shapeInfo());
+  auto rank = input.rankOf();
+  if (rank > axis) {
+    if (input.ordering() == 'c') {
+      // check befor axis
+      for (int i = 0; i < axis; i++) {
```
Comment (agibsonccc) [inline libnd4j/include/ops/impl/specials_single.hpp:144]: typo

**Review by agibsonccc — CHANGES_REQUESTED** [review]

File: libnd4j/include/ops/impl/specials_single.hpp (hunk 7 above)
Comment (quickwritereader) [reply libnd4j/include/ops/impl/specials_single.hpp]: I will remove it at all as it should be guaranteed already

File: libnd4j/include/ops/impl/specials_single.hpp (hunk 6 above)
Comment (quickwritereader) [reply libnd4j/include/ops/impl/specials_single.hpp:248]: its ok here

**Review by agibsonccc — APPROVED** [review]

**Pull request merged**